pareto/utilities/process_data.py

```diff
@@ -0,0 +1,838 @@
+                f"{t} ({round(strategic_model.p_gamma_TimePeriodDemand[t].value)} {pyunits.get_units(strategic_model.p_gamma_TimePeriodDemand)}s demand)"
+            )
+
+    if len(capacity_feasibility_message) > 0:
+        error_message = ", ".join(capacity_feasibility_message)
+        raise DataInfeasibilityError(
+            "An infeasibility in the input data has been detected. The following time periods have larger volumes of produced water than capacity in the system: "
+            + error_message
+        )
+
+    if len(demand_feasibility_message) > 0:
+        error_message = ", ".join(demand_feasibility_message)
+        raise DataInfeasibilityError(
+            "An infeasibility in the input data has been detected. The following time periods have higher demand than volume of produced water and externally sourced water available: "
+            + error_message
+        )
+
+    return strategic_model
+
+
+def _get_max_value_for_parameter(parameter):
+    return max([x.value for x in parameter.values()])
+
+
+# Custom error for Missing Data.
+class MissingDataError(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+
+# Custom error for Data Infeasibility
+class DataInfeasibilityError(Exception):
+    def __init__(self, message):
+        super().__init__(message)
+
+
+# Helper function to check for data that is expected if an optional set
+# (e.g. node type like "ReuseOptions") is given
+def _check_optional_data(
+    df_sets,
+    df_parameters,
+    optional_set_name,  # e.g., "ReuseOptions"
+    required_sets_with_option,  # []
+    required_parameters_with_option,  # ["BeneficialReuseCost","BeneficialReuseCredit"]
+    requires_at_least_one=[],  # ["ROA", "SOA", "NOA", "ROT", "SOT"]
+):
+    # create set object for df_sets and df_parameters for simpler list comparison
+    _df_sets_set = set(df_sets)
+    _df_parameters_set = set(df_parameters.keys())
+
+    # Check that optional_set_name is given by user (if not, no need to check)
+    if optional_set_name in _df_sets_set:
+        # Get missing parameters and sets
+        _missing_parameters = set(required_parameters_with_option) - _df_parameters_set
+        _missing_sets = set(required_sets_with_option) - _df_sets_set
+        # For each missing parameter, add an empty dictionary to df_parameters and raise a warning
+        if len(_missing_parameters) > 0 or len(_missing_sets) > 0:
+            for param in _missing_parameters:
+                df_parameters[param] = {}
+            for s in _missing_sets:
+                df_sets[s] = {}
+            _missing_tabs = _missing_sets
+            _missing_tabs.update(_missing_parameters)
+            warning_message = (
+                f"{optional_set_name} are given, but {optional_set_name} data is missing. The following tabs have been set to default values:"
+                + str(_missing_tabs)
+            )
+            warnings.warn(warning_message)
```

Comment:
I believe adding a stacklevel=3 would be better to avoid warnings.warn(warning_message) being printed.

Reply:
I like it!